Added type hints for pygame

MightyJosip opened the pull request:
Take note: this branch auto enables type hints, If you want I can disable them again. Here I covered (almost) all classes/functions that are mentioned in docs. Relevant issue #839. Work left to do:

- Add https://www.python.org/dev/peps/pep-0544/ to some functions (right now only on python3.8): This would be useful for example in pygame.sprite.collide_mask() where it would accept any Sprite that has the attribute mask
- Add pygame.Overlay, pygame.cdrom (I don't have this modules installed with pygame,)
- Do testing

MightyJosip commented:
Version 0.2:
- added pygame.mixer.music
- added pygame._sdl2.touch
- added pygame.display.Info() function and VidInfo class
- disabled pygame.sprite.LayeredDirty from accepting regular Sprite
- added support for numpy.ndarray object
- fixed bug where pygame.Surface methods wouldn't accept pygame.Color argument

MyreMylar commented:
This sounds very promising, I'm hopeful this will eventually also resolve #1009 by magic.

MightyJosip commented:
Version 0.3 (In this version I added support for mypy):
- Added stubs for pygame.constants (because otherwise mypy doesn't accept constants like pygame.KEYDOWN or pygame.QUIT)
- Moved _sdl2 to the another folder (now type checker accepts touch as module instead of variable)
- Modified setup.py to add .pyi files
- Various bug fixes

MightyJosip commented:
Event was mostly done by someone else (there were a few files that I already found when I started this), but yes it would be great if we have same keywords in docs/type hints

illume commented on buildconfig/pygame-stubs/__init__.pyi:
```diff
@@ -1,14 +1,59 @@
+def register_quit(callable: Callable) -> None: ...
 
-def __getattr__(name) -> Any: ...  # don't error on missing stubs
+# def __getattr__(name) -> Any: ...  # don't error on missing stubs
```
Is this commented out for debugging, or are there now no missing stubs?

MightyJosip replied:
Yes thanks I forgot to uncomment this, this should be almost all stubs (missing Overlay and cdrom (which are MissingModule on my pygame version) and some undocumented stuff (like pygame.HAVE_NEWBUF)

illume commented on buildconfig/pygame-stubs/bufferproxy.pyi:
```diff
@@ -1,11 +1,12 @@
     parent: Any
     length: int
-    raw: bytes
+    raw: str
```
This is bytes in python3 and str in python2. pgcompat.h

MightyJosip replied:
I had some problems with mypy where it couldn't accept it as AnyStr, but I will rewrite it as Union[bytes, str]

illume commented on buildconfig/pygame-stubs/camera.pyi:
```diff
@@ -0,0 +1,17 @@
+    def start(self) -> None: ...
+    def stop(self) -> None: ...
+    def get_controls(self) -> Tuple[bool, bool, int]: ...
+    def set_controls(self, hflip: bool, vflip: bool, brightness: int) -> Tuple[bool, bool, int]: ...
```
I think each of these arguments are optional. So this is possible: `set_controls(hflip=True)`. I'm not sure if this type captures that?

MightyJosip replied:
Yes you are right, I missed that part from docs

illume commented on buildconfig/pygame-stubs/color.pyi:
```diff
@@ -18,3 +18,4 @@ class Color(object):
     def normalize(self) -> Tuple[float, float, float, float]: ...
     def correct_gamma(self, gamma: float) -> Color: ...
     def set_length(self, len: int) -> None: ...
+    def lerp(self, color: Color, value: float) -> Color: ...
```
Is a type in the range 0.0 to 1.0 possible?
The keywords are color, and amount according to color.c:861
`    static char *keywords[] = {"color", "amount", NULL};`

MightyJosip replied:
I changed keyword, and for range I will try https://github.com/python/typing/issues/554

illume commented on buildconfig/pygame-stubs/__init__.pyi:
```diff
@@ -1,14 +1,59 @@
+import pygame.scrap
+
+# This classes are auto imported with pygame, so I put their declaration here
+class Rect(pygame.rect.Rect): ...
```
These aren't technically classes.
I wonder why this has been done, instead of the previous approach?
```python
Color = pygame.color.Color
```

MightyJosip replied:
For example with Rect = pygame.rect.Rect if I use pygame.Rect((20, 20, 20, 20)) in my program PyCharm will give me warning saying parameters y, width, height unfilled. Meanwhile mypy will not find any warning (but there are a few otherproblems with PyCharm overloading) so I can rewrite them like they used to be

illume commented on buildconfig/pygame-stubs/cursors.pyi:
```diff
@@ -0,0 +1,23 @@
+from typing import Tuple, Sequence, Optional
+
+_Bitmap = Tuple[Tuple[int, int], Tuple[int, int],
```
This is my favourite type signature. hehe. 😻

MightyJosip replied:
Hahaha my inspiration was issue 221 in python/typing (didn't put link because it looks like it gets mentioned even if I write link in other projects, don't want to overwhelm them with updates :))

illume commented on buildconfig/pygame-stubs/constants.pyi:
```diff
@@ -0,0 +1,521 @@
+'''
```
I guess there will be some constant differences depending on platform, python version, and SDL version.

MightyJosip replied:
If that is the case than it could maybe cause some problems with mypy, because they don't want something like from pygame.constants import *; they want either pygame.constants to have type hints like this or to have __all__ (which would basically mean the same thing). Btw without this file it worked well in PyCharm (because is is able to see the constants file).

illume commented on buildconfig/pygame-stubs/__init__.pyi:
```diff
@@ -1,14 +1,59 @@
+def set_error(error_msg: str) -> None: ...
+def get_sdl_version() -> Tuple[int, int, int]: ...
+def get_sdl_byteorder() -> int: ...
+def encode_string(obj: Union[str, bytes], encoding: Optional[str]='unicode_escape', errors: Optional[str]='backslashreplace', etype: Optional[Type[Exception]]=UnicodeEncodeError) -> bytes: ...
```
I guess these long lines can be broken. Do you know if the `black` python formatter works with type stubs or not?

MightyJosip replied:
Done

illume commented on buildconfig/pygame-stubs/_sdl2/touch.pyi:
```diff
@@ -0,0 +1,7 @@
+
+def get_num_devices() -> int: ...
+def get_device(index: int) -> int: ...
+def get_num_fingers(touch_id: int) -> int: ...
```
`touch.c@pg_touch_num_fingers` does not use a `touch_id` keyword for get_num_fingers.

MightyJosip replied:
If this is device_id that I fixed it

illume commented on buildconfig/pygame-stubs/display.pyi:
```diff
@@ -0,0 +1,57 @@
+def get_driver() -> str: ...
+def Info() -> _VidInfo: ...
+def get_wm_info() -> Dict[str, int]: ...
+def list_modes(depth: Optional[int]=0, flags: Optional[int]=-2147483648, display: Optional[int]=0) -> List[Tuple[int, int]]: ...
```
This magic number is a bit weird. Is it PGS_FULLSCREEN? I wonder what this number is on 32bit vs 64bit systems.

MightyJosip replied:
Fixed

illume commented on buildconfig/pygame-stubs/font.pyi:
```diff
@@ -0,0 +1,29 @@
+    def get_underline(self) -> bool: ...
+    def set_bold(self, value: bool) -> None: ...
+    def get_bold(self) -> bool: ...
+    def set_italic(self, value: bool) -> None: ...
```
font.c@font_set_italic does not use a keyword argument.

MightyJosip replied:
Had to write something

illume commented on buildconfig/pygame-stubs/gfxdraw.pyi:
```diff
@@ -0,0 +1,33 @@
+def box(surface: Surface, rect: _RectValue, color: _ColorValue) -> None: ...
+def circle(surface: Surface, x: int, y: int, r: int, color: _ColorValue) -> None: ...
+def aacircle(surface: Surface, x: int, y: int, r: int, color: _ColorValue) -> None: ...
+def filled_circle(surface: Surface, x: int, y: int, r: int, color: _ColorValue) -> None: ...
```
`grxdraw.c@_gfx_filledcirclecolor` does not use keyword arguments.

illume commented on buildconfig/pygame-stubs/image.pyi:
```diff
@@ -0,0 +1,11 @@
+def save(surface: Surface, filename: str) -> None: ...
+def get_extended() -> bool: ...
+def tostring(surface: Surface, format: str, flipped: Optional[bool]=False) -> str: ...
+def fromstring(string: str, size: Union[List[int], Tuple[int, int]], format: str, flipped: Optional[bool]=False) -> Surface: ...
```
image.c@image_fromstring does not use keyword arguments.
Is there a way to specify list of only 2 elements [int, int]?

MightyJosip replied:
Sadly not (or at least I didn't find any way): https://stackoverflow.com/questions/44833822/specify-length-of-sequence-or-list-with-python-typing-module. That is why I put both list and tuple, and not for example sequence (so users will at least get warning if they use tuple).

illume commented on buildconfig/pygame-stubs/key.pyi:
```diff
@@ -0,0 +1,18 @@
+def get_pressed() -> Sequence[bool]: ...
+def get_mods() -> int: ...
+def set_mods() -> int: ...
+def set_repeat(delay: Optional[int], interval: Optional[int]) -> None: ...
```
key.c@key_set_repeat does not use keyword arguments.

illume commented on buildconfig/pygame-stubs/mask.pyi:
```diff
@@ -0,0 +1,40 @@
+    def __init__(self, size: Union[List[int], Tuple[int, int]], fiil: Optional[bool]=False) -> None: ...
+    def copy(self) -> Mask: ...
+    def get_size(self) -> Tuple[int, int]: ...
+    def get_rect(self, **kwargs) -> Rect: ...  # Dict type needs to be completed
```
I guess this note could be marked with a TODO, so it's easier to find.

MightyJosip replied:
I have TODO file on my PC, but I can add it somewhere if you want

illume commented on buildconfig/pygame-stubs/mask.pyi:
```diff
@@ -0,0 +1,40 @@
+
+
+def from_surface(surface: Surface, threshold: Optional[int]=127) -> Mask: ...
+def from_threshold(surface: Surface, color: _ColorValue, threshold: Optional[_ColorValue]=(0, 0, 0, 255), other_surface: Optional[Surface]=None, palette_colors: Optional[int]=1) -> Mask: ...
```
mask.c@mask_from_threshold does not use keyword arguments.

illume commented on buildconfig/pygame-stubs/mask.pyi:
```diff
@@ -0,0 +1,40 @@
+    def get_at(self, pos: Union[List[int], Tuple[int, int]]) -> int: ...
+    def set_at(self, pos: Union[List[int], Tuple[int, int]], value: Optional[int]=1) -> None: ...
+    def overlap(self, othermask: Mask, offset: Union[List[int], Tuple[int, int]]) -> Union[Tuple[int, int], None]: ...
+    def overlap_area(self, othermask: Mask, offset: Union[List[int], Tuple[int, int]]) -> int: ...
```
I guess offset can be any sequence of 2 ints, not just lists, and tuple?

MightyJosip replied:
I can add it, but then it would accept tuple with any number of ints (because you can't check length of regular sequence)

illume commented on buildconfig/pygame-stubs/math.pyi:
```diff
@@ -0,0 +1,84 @@
+    def angle_to(self, other: Vector2) -> float: ...
+    def as_polar(self) -> Tuple[float, float]: ...
+    def from_polar(self, polar_value: Union[List[float], Tuple[float, float]]) -> None: ...
+    def update(self, first: Optional[Union[float, Vector2, Tuple[float, float], List[float]]]=0, second: Optional[float]=0) -> None: ...
```
vector.c@vector2_update has keyword arguments of `x, y` not `first` and `second`.

MightyJosip replied:
Done

illume commented on buildconfig/pygame-stubs/math.pyi:
```diff
@@ -0,0 +1,84 @@
+    def scale_to_length(self, value: float) -> None: ...
+    def reflect(self, other: Vector2) -> Vector2: ...
+    def reflect_ip(self, other: Vector2) -> None: ...
+    def distance_to(self, other: Vector2) -> float: ...
```
other can be any sequence, not just Vector2. See vector.c@_vector_distance_helper

MightyJosip replied:
Tnx

illume commented on buildconfig/pygame-stubs/mouse.pyi:
```diff
@@ -0,0 +1,15 @@
+def get_pressed() -> Tuple[int, int, int]: ...
+def get_pos() -> Tuple[int, int]: ...
+def get_rel() -> Tuple[int, int]: ...
+@overload
```
👍 TIL: `@overload`. Cool.

illume commented on buildconfig/pygame-stubs/mouse.pyi:
```diff
@@ -0,0 +1,15 @@
+def set_visible(value: bool) -> int: ...
+def get_visible() -> bool: ...
+def get_focused() -> int: ...
+def set_cursor(size: Union[Tuple[int, int], List[int]], hotspot: Union[Tuple[int, int], List[int]], xormasks: Sequence[int], andmasks: Sequence[int]) -> None: ... # This needs to be checked
```
Probably needs a TODO note to check.

illume commented on buildconfig/pygame-stubs/pixelarray.pyi:
```diff
@@ -0,0 +1,20 @@
+    strides: Tuple[int, ...]
+    def __init__(self, surface: Surface) -> None: ...
+    def make_surface(self) -> Surface: ...
+    def replace(self, color: _ColorValue, repcolor: _ColorValue, distance: Optional[float]=0, weights: Optional[Tuple[float, float, float]]=(0.299, 0.587, 0.114)) -> None: ...
```
Would be easier to read with shorter lines.

`weights` can be any sequence with 3 elements. See pixelarray_methods.c@_get_weights

illume commented on buildconfig/pygame-stubs/pixelcopy.pyi:
```diff
@@ -0,0 +1,9 @@
+
+from pygame.surface import Surface
+
+def surface_to_array(array: numpy.ndarray, surface: Surface, kind: Optional[str]='P', opaque: Optional[int]=255, clear: Optional[int]=0) -> None: ...
```
`array` here seems to be anything that `pgObject_GetBuffer` works on, not just `numpy.ndarray`.

illume replied:
`kind` can be one of these different characters.

From the docs:
> Recognized kind values are the single character codes 'P', 'R', 'G', 'B', 'A', and 'C'. Kind codes are case insensitive, so 'p' is equivalent to 'P'.

MightyJosip replied:
For now I can replace it with Any, to get if something works on something I would have to use typing.Protocol, maybe in future I will add it, Kind is fixed

illume commented on buildconfig/pygame-stubs/transform.pyi:
```diff
@@ -0,0 +1,26 @@
+def rotate(surface: Surface, angle: float) -> Surface: ...
+def rotozoom(surface: Surface, angle: float, scale: float) -> Surface: ...
+def scale2x(surface: Surface, dest_surface: Optional[Surface]=None) -> Surface: ...
+def smoothscale(surface: Surface, size: Union[Tuple[int, int], List[int]], dest_surface: Optional[Surface]=None) -> Surface: ...
```
transform.c@surf_scalesmooth doesn't use keyword arguments.

illume commented on buildconfig/pygame-stubs/time.pyi:
```diff
@@ -0,0 +1,13 @@
+def delay(milliseconds: int) -> int: ...
+def set_timer(eventid: Union[int, Event], milliseconds: int, once: Optional[bool]=False) -> None: ...
+class Clock:
+    def tick(self, framerate: Optional[int]=0) -> int: ...
```
time.c@clock_tick_base doesn't use keyword arguments.

illume commented on buildconfig/pygame-stubs/time.pyi:
```diff
@@ -0,0 +1,13 @@
+def get_ticks() -> int: ...
+def wait(milliseconds: int) -> int: ...
+def delay(milliseconds: int) -> int: ...
+def set_timer(eventid: Union[int, Event], milliseconds: int, once: Optional[bool]=False) -> None: ...
```
time.c@time_set_timer set_timer doesn't use keyword arguments.

I guess `eventid` should just be called `event`, or `event_type`.

illume commented on buildconfig/pygame-stubs/sprite.pyi:
```diff
@@ -0,0 +1,112 @@
+    def repaint_rect(self, screen_rect: _RectStyle) -> None: ...
+    def set_clip(self, screen_rect: Optional[_RectStyle]=None): ...
+    def get_clip(self) -> Rect: ...
+    def set_timing_treshold(self, time_ms: Any) -> None: ...  # This actually accept any value
```
You fixed this in https://github.com/pygame/pygame/pull/1688

MightyJosip replied:
Tnx.

illume commented on buildconfig/pygame-stubs/surface.pyi:
```diff
@@ -53,13 +59,18 @@ class Surface(object):
-    def set_palette(self, palette: List[_RgbaInput]) -> None: ...
-    def set_palette_at(self, index: int, color: _RgbaInput) -> None: ...
-    def map_rgb(self, color: _RgbaInput) -> int: ...
+    def set_palette(self, palette: List[_ColorInput]) -> None: ...
```
`set_palette` does not use keyword arguments, see surface.c@surf_set_palette
